frontend/middle-2/design.md

```diff
@@ -1,17 +1,48 @@
 #Базовые принципы проектирования
 
-* Что такое инкапсуляция? Какова мотивация для улучшения уровня инкапсуляции?
-* Что такое абстрагирование? Какие есть отличия между абстракцией, инкапсуляцией и скрытием информации?
-* Что такое coupling? Как инкапсуляция влияет на это?
-* Что такое cohesion? Является ли эта величина обратной величине coupling?
-* Что такое принцип DRY? Как принцип связан с cohesion?
-* Что такое принцип KISS?
-* Что такое принцип YAGNI? Как он соотносится с принципом KISS?
-* Что такое Separation of concerns?
-* Из-за чего возникает проблема impedance mismatch при обмене данными? Какие есть примеры возникновения проблемы в разных технологиях? Какие там используются решения?
-* Почему преждевременная оптимизация может легко обернуться проблемой, а не решением при проектировании систем?
-
+* Inversion of Control
+  * What is IoC?
+  * What are the advantages of frameworks providing IoC over libraries?
+  * What are the advantages of libraries without IoC over frameworks?
+  * Can we expose such module's interface that support both approaches: with IoC and without IoC?
+* Dependency Injection
+  * What is DI?
+  * What are the alternatives to DI? What are the advantages and disadvantages of each alternative?
+  * Can we implement DI without IoC?
+  * How does DI help us to achieve the Separation of Concerns?
+  * How are DI and Dependency Inversion Principle from SOLID related to each other?
+* What is the Impedance Mismatch? What are examples of Impedance Mismatch among different technologies? What are most popular patterns of solutions?
```

Comment:
What is Impedance Mismatch? What are the examples of Impedance Mismatch in different technologies? What are the most popular patterns of solutions?